src/middlewared/middlewared/plugins/cloud_sync.py

```diff
@@ -0,0 +1,522 @@
+        Str('dayweek'),
+        Str('month'),
+        Dict('attributes', additional_attrs=True),
+        Bool('enabled', default=True),
+        register=True,
+    ))
+    async def do_create(self, cloud_sync):
+        """
+        Creates a new cloud_sync entry.
+
+        .. examples(websocket)::
+
+          Create a new cloud_sync using amazon s3 attributes, which is supposed to run every hour.
+
+            :::javascript
+            {
+              "id": "6841f242-840a-11e6-a437-00e04d680384",
+              "msg": "method",
+              "method": "cloudsync.create",
+              "params": [{
+                "description": "s3 sync",
+                "path": "/mnt/tank",
+                "credentials": 1,
+                "minute": "00",
+                "hour": "*",
+                "daymonth": "*",
+                "month": "*",
+                "attributes": {
+                  "bucket": "mybucket",
+                  "folder": ""
+                },
+                "enabled": true
+              }]
+            }
+        """
+
+        verrors = ValidationErrors()
+
+        await self._validate(verrors, 'cloud_sync', cloud_sync)
+
+        if verrors:
+            raise verrors
+
+        await self._validate_folder(verrors, 'cloud_sync', cloud_sync)
+
+        if verrors:
+            raise verrors
+
+        cloud_sync = await self._compress(cloud_sync)
+
+        cloud_sync["id"] = await self.middleware.call('datastore.insert', 'tasks.cloudsync', cloud_sync)
+        await self.middleware.call('notifier.restart', 'cron')
+
+        cloud_sync = await self._extend(cloud_sync)
+        return cloud_sync
+
+    @accepts(Int('id'), Patch('cloud_sync', 'cloud_sync_update', ('attr', {'update': True})))
+    async def do_update(self, id, data):
+        """
+        Updates the cloud_sync entry `id` with `data`.
+        """
+        cloud_sync = await self._get_cloud_sync(id)
+
+        # credentials is a foreign key for now
+        if cloud_sync['credentials']:
+            cloud_sync['credentials'] = cloud_sync['credentials']['id']
+
+        cloud_sync.update(data)
+
+        verrors = ValidationErrors()
+
+        await self._validate(verrors, 'cloud_sync_update', cloud_sync)
+
+        if verrors:
+            raise verrors
+
+        await self._validate_folder(verrors, 'cloud_sync_update', cloud_sync)
+
+        if verrors:
+            raise verrors
+
+        cloud_sync = await self._compress(cloud_sync)
+
+        await self.middleware.call('datastore.update', 'tasks.cloudsync', id, cloud_sync)
+        await self.middleware.call('notifier.restart', 'cron')
+
+        cloud_sync = await self._extend(cloud_sync)
+        return cloud_sync
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        """
+        Deletes cloud_sync entry `id`.
+        """
+        await self.middleware.call('datastore.delete', 'tasks.cloudsync', id)
+        await self.middleware.call('notifier.restart', 'cron')
```

Comment:
Better if we use service.restart here I would think, one less thing to worry about when notifier goes away.